### PR TITLE
DVC 5112 store and read anonymousId

### PR DESCRIPTION
**Pull request by L-Joli**
- add store and read anonymous id logic to `Client`

### CODE REVIEW

File: sdk/js/src/Client.ts [hunk 1]
```diff
@@ -46,7 +46,20 @@ export class DVCClient implements Client {
     private windowMessageHandler?: (event: MessageEvent) => void
 
     constructor(environmentKey: string, user: DVCUser, options: DVCOptions = {}) {
+        this.logger = options.logger || dvcDefaultLogger({ level: options.logLevel })
+        this.store = new Store(typeof window !== 'undefined' ? window.localStorage : stubbedLocalStorage, this.logger)
+        
+        const storedAnonymousId = this.store.store?.getItem(StoreKey.AnonUser)
```
Comment (rodeora) [inline sdk/js/src/Client.ts:52]: it would be better to use `this.store.load` function instead

File: sdk/js/__tests__/Client.spec.js [hunk 2]
```diff
@@ -134,6 +135,24 @@ describe('DVCClient tests', () => {
         expect(client.config).toBeUndefined()
     })
 
+    it('should have same value for anonymous user id from local storage and user_id if anonId exists and isAnonymous is true', async () => {
+        const client = new DVCClient('test_env_key', { isAnonymous: true })
+        const anonymousUserId = client.store.store.getItem(StoreKey.AnonUser)
```
Comment (rodeora) [inline sdk/js/__tests__/Client.spec.js:140]: use client.store.load instead

File: sdk/js/__tests__/Client.spec.js [hunk 3]
```diff
@@ -134,6 +135,24 @@ describe('DVCClient tests', () => {
...
+
+    it('should save anonymous user id in local storage if isAnonymous is true', async () => {
+        const client = new DVCClient('test_env_key', { isAnonymous: true })
+        const anonymousUserId = client.store.store.getItem(StoreKey.AnonUser)
+        expect(anonymousUserId).toBeTruthy()
+    })
+
+    it('should not save anonymous user id in local storage if isAnonymous is false', async () => {
+        const client = new DVCClient('test_env_key', { user_id: 'user1', isAnonymous: false })
+        const anonymousUserId = client.store.store.getItem(StoreKey.AnonUser)
```
Comment (rodeora) [inline sdk/js/__tests__/Client.spec.js:152]: same here

File: sdk/js/__tests__/Client.spec.js [hunk 4]
```diff
@@ -134,6 +135,24 @@ describe('DVCClient tests', () => {
...
+    it('should have same value for anonymous user id from local storage and user_id if anonId exists and isAnonymous is true', async () => {
+        const client = new DVCClient('test_env_key', { isAnonymous: true })
+        const anonymousUserId = client.store.store.getItem(StoreKey.AnonUser)
+        expect(JSON.parse(anonymousUserId)).toEqual(client.user.user_id)
+    })
+
+    it('should save anonymous user id in local storage if isAnonymous is true', async () => {
+        const client = new DVCClient('test_env_key', { isAnonymous: true })
+        const anonymousUserId = client.store.store.getItem(StoreKey.AnonUser)
+        expect(anonymousUserId).toBeTruthy()
```
Comment (rodeora) [inline sdk/js/__tests__/Client.spec.js:147]: this is testing the same thing as the previous test, you can delete this

**Review by rodeora — CHANGES_REQUESTED** [review]

File: sdk/js/src/Client.ts (hunk 1 above)
Comment (L-Joli) [reply sdk/js/src/Client.ts:52]: ummm...but it returns a Promise, do we really need a Promise for this case?

Comment (L-Joli) [reply sdk/js/src/Client.ts]: I can clear all the promises in store in other pr

File: sdk/js/src/Store.ts [hunk 5]
```diff
@@ -22,10 +22,8 @@ export class Store {
         }, storeKey)
     }
 
-    load(storeKey: string): Promise<string | null | undefined> {
-        return new Promise((resolve) => {
-            resolve(this.store?.getItem(storeKey))
-        })
+    load(storeKey: string): string | null | undefined{
```
Comment (rodeora) [inline sdk/js/src/Store.ts:25]: can you un-promisify everything in here?

**Review by rodeora — CHANGES_REQUESTED** [review]

File: sdk/js/src/Store.ts [hunk 6]
```diff
@@ -16,53 +16,38 @@ export class Store {
...
+            resolve(this.save(StoreKey.Config, data))
+        })
     }
 
-    loadConfig(): Promise<string | null | undefined> {
+    loadConfig(): string | null | undefined {
         return this.load(StoreKey.Config)
     }
 
     saveUser(user: DVCPopulatedUser): Promise<void> {
```
Comment (rodeora) [inline sdk/js/src/Store.ts:41]: this shouldn't be a promise either

File: sdk/js/src/Store.ts [hunk 7]
```diff
@@ -16,53 +16,38 @@ export class Store {
...
-            })
-        } catch (e) {
-            this.logger?.error(e as string)
-            return Promise.resolve()
-        }
+    remove(storeKey: string): void {
+        this.store?.removeItem(storeKey)
     }
 
     saveConfig(data: BucketedUserConfig): Promise<void> {
```
Comment (rodeora) [inline sdk/js/src/Store.ts:31]: remove the promise here

**Review by rodeora — CHANGES_REQUESTED** [review]

File: sdk/js/src/Store.ts (hunk 6 above)
Comment (L-Joli) [reply sdk/js/src/Store.ts:41]: but in `Client.ts`, seems like we want to log the success message. If it's not a promise, we don't know if it's succeeded. Do we want to keep the logger and promise or just remove both? same for the `saveConfig` case
```
this.store.saveUser(user).then(() => {
       this.logger.info('Successfully saved user to local storage')
})
```

Comment (rodeora) [reply sdk/js/src/Store.ts:41]: you can just log it here instead of the client

File: sdk/js/src/Store.ts [hunk 8]
```diff
@@ -16,53 +16,33 @@ export class Store {
...
-
-    loadConfig(): Promise<string | null | undefined> {
+    loadConfig(): string | null | undefined {
         return this.load(StoreKey.Config)
     }
 
-    saveUser(user: DVCPopulatedUser): Promise<void> {
-        if (!user) {
-            return Promise.reject('No user to save')
-        }
```
Comment (rodeora) [inline sdk/js/src/Store.ts:57]: still want to keep this, throw an error

**Review by rodeora — CHANGES_REQUESTED** [review]

File: sdk/js/__tests__/Client.spec.js [hunk 9]
```diff
@@ -134,6 +135,18 @@ describe('DVCClient tests', () => {
...
+        const anonymousUserId = client.store.load(StoreKey.AnonUser)
+        expect(JSON.parse(anonymousUserId)).toEqual(client.user.user_id)
+    })
+
+    it('should not save anonymous user id in local storage if isAnonymous is false', async () => {
+        const client = new DVCClient('test_env_key', { user_id: 'user1', isAnonymous: false })
+        const anonymousUserId = client.store.load(StoreKey.AnonUser)
+        expect(anonymousUserId).toBeNull()
+    })
+
```
Comment (rodeora) [inline sdk/js/__tests__/Client.spec.js:149]: we need a test to check an anon_user gets the same id as the one in local storage

**Review by rodeora — CHANGES_REQUESTED** [review]

File: sdk/js/__tests__/Client.spec.js (hunk 9 above)
Comment (L-Joli) [reply sdk/js/__tests__/Client.spec.js:149]: umm... this test `should have same value for anonymous user id from local storage and user_id if anonId exists and isAnonymous is true` includes the checking "anonymous user id =  the one in local storage". Is there something else I need to add into this?

Comment (rodeora) [reply sdk/js/__tests__/Client.spec.js:149]: you need a new test, the existing one is checking that the anon user's id was saved to local storage. The new test needs to check that a new anon user's id gets set to the same value as the one saved in local storage

**Review by rodeora — APPROVED** [review]

**Pull request merged**